pravega/src/main/scala/akka/stream/alpakka/pravega/impl/PravegaSource.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.pravega.impl
+
+import akka.stream.stage.{GraphStageLogic, GraphStageWithMaterializedValue, OutHandler, StageLogging}
+import akka.stream.{Attributes, Outlet, SourceShape}
+import akka.Done
+import akka.annotation.InternalApi
+import akka.event.Logging
+import akka.stream.alpakka.pravega.ReaderSettings
+import io.pravega.client.ClientConfig
+
+import scala.concurrent.{Future, Promise}
+import io.pravega.client.stream.{EventStreamReader, ReaderGroup}
+
+import scala.util.control.NonFatal
+import akka.stream.ActorAttributes
+@InternalApi private final class PravegaSourcesStageLogic[A](
+    shape: SourceShape[A],
+    val scope: String,
+    streamName: String,
+    val readerSettings: ReaderSettings[A],
+    startupPromise: Promise[Done]
+) extends GraphStageLogic(shape)
+    with PravegaReader
+    with StageLogging {
+
+  override protected def logSource = classOf[PravegaSourcesStageLogic[A]]
+
+  private def out = shape.out
+
+  private var reader: Reader[A] = _
+
+  override def clientConfig: ClientConfig = readerSettings.clientConfig
+
+  setHandler(
+    out,
+    new OutHandler {
+
+      override def onDownstreamFinish(): Unit = {
+        log.debug("Downsteam finished")
+        super.onDownstreamFinish()
+      }
+
+      override def onPull(): Unit = {
+        val eventRead = reader.nextEvent(readerSettings.timeout)
+        if (eventRead == null)
+          log.debug("timeout")
+        else {
+          val event = eventRead.getEvent
```

Comment:
I'm thinking it may make sense to propagate more of the `ReaderEvent` metadata downstream, such as the offset and segment id.  This information would be useful for app-level consistency guarantees during stream startup.  If we create a wrapper type now that includes this metadata then it won't be a breaking change later if someone needs it.

Reply:
Agree, should I consider this approach https://blog.softwaremill.com/painlessly-passing-message-context-through-akka-streams-1615b11efc2c

Reply:
Indeed, using `FlowWithContext` and `SourceWithContext` would be great.  They do have limitations though.  Not all stage operations are available and there are situations where mapping the context is not always obvious (such as when you explode out a source context in `mapConcat`, I'm actually working on a proposal to make this easier in Akka Streams!).  However, users can always drop down to normal `Flow` and `Source` which will then emit the context and element are mapped as a tuple.  

In Alpakka Kafka we expose both APIs, but since this is a net new connector I think it would be cool to use the withContext APIs from the start if you're interested in exploring that API.  There's not a lot of docs around it.  In fact, I wasn't even aware of that software mill blog post (thanks!)  So just ask if you have any questions.  If you go this path then we should make a note about their usage in the docs, such as how to switch back and forth between them.

Reply:
We can tackle with context APIs in a different PR.